### PR TITLE
Feature/refactor config

### PR DESCRIPTION
**Pull request by martonborzak**
Added 2 methods to extend the YIO API functionalities. These methods will help the web configurator to create a drag and drop management interface for entities.

`getAllAvailableEntities` returns a list of available entities. These entities are not controlled by the remote, but available from the integration.

`addAvailableEntity` adds an entity to the list of available entities. It checks if the entity type is supported. If so, it returns true, otherwise false. This is implemented in `integration` class and can be used by all integration implementations.

### CODE REVIEW

**Comment by zehnm** [comment]
Please provide a description what this pull request contains and the purpose of the changes.
E.g. what is the reason to change `getAllAvailableEntities()`to now return a `QVariantList` instead of `QStringList`?
This is a breaking change which affects all integration plugins, but there are no pull requests yet.

File: src/yio-interface/entities/entitiesinterface.h [hunk 1]
```diff
@@ -48,6 +48,9 @@ class EntitiesInterface {
 
     // get entity interface by entity_id
     virtual EntityInterface* getEntityInterface(const QString& entity_id) = 0;
+
+    // get a list of supported entities
+    virtual QStringList supported_entities() = 0;
```
Comment (zehnm) [inline src/yio-interface/entities/entitiesinterface.h]: Use lowerCamelCase for naming: `supportedEntities`

File: src/yio-interface/integrationinterface.h [hunk 2]
```diff
@@ -57,7 +57,7 @@ class IntegrationInterface {
     /**
      * @brief getAllAvailableEntities Can be implemented by integration
      */
-    virtual QStringList getAllAvailableEntities() = 0;
+    virtual QVariantList getAllAvailableEntities() = 0;
```
Comment (zehnm) [inline src/yio-interface/integrationinterface.h]: Why is this change required?
What are the actual types that can be returned? E.g. does a list of integers make sense?

File: src/yio-plugin/integration.h [hunk 3]
```diff
@@ -100,5 +100,9 @@ class Integration : public QObject, public IntegrationInterface {
     YioAPIInterface*        m_yioapi;
     ConfigInterface*        m_config;
     QLoggingCategory&       m_logCategory;
-    QStringList             m_allAvailableEntities;
+    QVariantList            m_allAvailableEntities;
+    /// containts QVariantMap objects with the following keys
+    /// "entity_id": entity id
+    /// "type": type of entity, like light
+    /// "integration_id": id of the integration controlling the entity
```
Comment (zehnm) [inline src/yio-plugin/integration.h]: I don't understand those comments. Are they related to the content of QVariantList?
Then the description must be added to the getter method in IntegrationInterface to describe the API.

File: src/yio-interface/integrationinterface.h [hunk 4]
```diff
@@ -57,7 +57,18 @@ class IntegrationInterface {
...
+
+    /**
+     * @brief sendCommand Must be implemented as Q_INVOKABLE
```
Comment (zehnm) [inline src/yio-interface/integrationinterface.h]: sendCommand?

File: src/yio-interface/integrationinterface.h [hunk 5]
```diff
@@ -57,7 +57,18 @@ class IntegrationInterface {
...
+     * @param entity_id
+     * @param type
+     * @param integration
+     * @param friendly_name
+     * @param supported_features
+     */
+    virtual bool addAvailableEntity(const QString& entity_id, const QString& type, const QString& integration,
+                                    const QString& friendly_name, const QStringList& supported_features) = 0;
```
Comment (zehnm) [inline src/yio-interface/integrationinterface.h]: please use lowerCamelCase for parameter naming

File: src/yio-plugin/integration.cpp [hunk 6]
```diff
@@ -69,6 +69,30 @@ Integration::Integration(Plugin* plugin)
 
 Integration::~Integration() {}
 
+bool Integration::addAvailableEntity(const QString& entity_id, const QString& type, const QString& integration,
+                                     const QString& friendly_name, const QStringList& supported_features) {
```
Comment (zehnm) [inline src/yio-plugin/integration.cpp]: lowerCamelCase for parameters as in header file

File: src/yio-plugin/integration.h [hunk 7]
```diff
@@ -59,11 +59,13 @@ class Integration : public QObject, public IntegrationInterface {
...
-    void             enterStandby() {}                                             // Can be overriden by integration
-    void             leaveStandby() {}                                             // Can be overriden by integration
-    QStringList      getAllAvailableEntities() { return m_allAvailableEntities; }  // Can be overriden by integration
+    Q_INVOKABLE void connect()    = 0;  // Must be implemented by integration
+    Q_INVOKABLE void disconnect() = 0;  // Must be implemented by integration
+    void             enterStandby() {}  // Can be overriden by integration
+    void             leaveStandby() {}  // Can be overriden by integration
+    QVariantList     getAllAvailableEntities() { return m_allAvailableEntities; }
+    bool             addAvailableEntity(const QString& entity_id, const QString& type, const QString& integration,
+                                        const QString& friendly_name, const QStringList& supported_features);
```
Comment (zehnm) [inline src/yio-plugin/integration.h]: lowerCamelCase naming

File: src/yio-interface/integrationinterface.h (hunk 4 above)
Comment (zehnm) [reply src/yio-interface/integrationinterface.h]: Please fix documentation, looks like copy pasted.

File: src/yio-plugin/integration_threadadapter.h [hunk 8]
```diff
@@ -51,16 +51,16 @@ class IntegrationThreadAdapter : public Integration {
     Q_INVOKABLE void disconnect() override;
     void             enterStandby() override;
     void             leaveStandby() override;
-    QStringList      getAllAvailableEntities() override;
-    Q_INVOKABLE void sendCommand(const QString& type, const QString& entity_id, int command,
+    QVariantList     getAllAvailableEntities() override;
+    Q_INVOKABLE void sendCommand(const QString& type, const QString& entityId, int command,
```
Comment (zehnm) [inline src/yio-plugin/integration_threadadapter.h]: I don't think we need a Q_INVOKABLE here. It's an internal helper class.
In my opinion all Q_INVOKABLE should be slots including enterStandby and leaveStandby.

File: src/yio-interface/integrationinterface.h [hunk 9]
```diff
@@ -56,23 +56,36 @@ class IntegrationInterface {
...
-     * @param entity_id
+     * @param integration
+     * @param friendlyName
+     * @param supportedFeatures
+     */
+    virtual bool addAvailableEntity(const QString& entityId, const QString& type, const QString& integration,
+                                    const QString& friendlyName, const QStringList& supportedFeatures) = 0;
+
+    /**
+     * @brief sendCommand Must be implemented as Q_INVOKABLE
```
Comment (zehnm) [inline src/yio-interface/integrationinterface.h]: Why does it need to be a Q_INVOKABLE?
Is there any QML dependency to the integration interface?

File: src/yio-plugin/integration.h [hunk 10]
```diff
@@ -55,31 +55,34 @@ class Integration : public QObject, public IntegrationInterface {
...
     Q_PROPERTY(int state READ state NOTIFY stateChanged)
 
-    Q_INVOKABLE void connect()    = 0;                                             // Must be implemented by integration
-    Q_INVOKABLE void disconnect() = 0;                                             // Must be implemented by integration
-    void             enterStandby() {}                                             // Can be overriden by integration
-    void             leaveStandby() {}                                             // Can be overriden by integration
-    QStringList      getAllAvailableEntities() { return m_allAvailableEntities; }  // Can be overriden by integration
-    Q_INVOKABLE void sendCommand(const QString& type, const QString& entity_id, int command, const QVariant& param) = 0;
+    Q_INVOKABLE void connect() override    = 0;  // Must be implemented by integration
+    Q_INVOKABLE void disconnect() override = 0;  // Must be implemented by integration
```
Comment (zehnm) [inline src/yio-plugin/integration.h]: I don't see the reason for Q_INVOKABLE. Everything should be handled by C++ now afaik. Therefore defining them as slots would be better, or I'm still missing something.
Or do we really allow people to write integration plugins in QML?

**Review by zehnm — COMMENTED** [review]
We need to go over the interfaces again, especially Q_INVOKABLE methods.
It would be nice to have a rather stable v1 version and not to break it again with the next release.

File: src/yio-plugin/integration.h (hunk 10 above)
Comment (martonborzak) [reply src/yio-plugin/integration.h]: connect and disconnect is `Q_INVOKABLE`, so you can manually disconnect the integration from the UI. If you go to `Settings/Integrations` there is option to connect/disconnect.

**Review by zehnm — APPROVED** [review]
As discussed we continue cleaning up the interfaces and create a minor release now to continue with the web configurator.
